Documentation/proposals/202405-agent-daemonset.md

```diff
@@ -0,0 +1,132 @@
+
+The reason for enhancing existing CRD (instead of introducing a new CRD) is it would take less time to finish the MVP. We’ll let users experiment with the MVP, and in case users report a separate CRD is needed, we’ll separate the logic of DaemonSet deployment into a new CRD later.
+
+The current [Prometheus Agent CRD](https://prometheus-operator.dev/docs/operator/api/#monitoring.coreos.com/v1alpha1.PrometheusAgent) already has sufficient fields for the DaemonSet deployment. The DaemonSet deployment can use all the existing fields in the CRD except the ones related to:
+* Selectors for service, probe, ScrapeConfig
+* Replica
+* Shard
+* Storage
+We can add a new bool field `daemonset`, which is defaulted to `false`. If the DaemonSet deployment is activated (`daemonset: true`), all the fields except the unrelated ones listed above will be used for the DaemonSet mode.
+
+### 6.2. Node detecting:
+
+As pointed out in [Danny from GMP’s talk](https://www.youtube.com/watch?v=yk2aaAyxgKw), to make Prometheus Agent DaemonSet know which node it’s on, we can use [Kubernetes’ downward API](https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/). In `prometheus` or `config-reloader` container, we can mount the node name as an environment variable like this:
+containers:
+```
+ - name: prometheus
+   image: prom/prometheus
+   env:
+   - name: NODE_NAME
+     valueFrom:
+     	fieldRef:
+     		apiVersion: v1
+     		fieldPath: spec.nodeName
+```
+
+### 6.3. Targets filtering for pods (PodMonitor support):
+
+To filter targets, Danny’s talk has pointed out two options.
+The first option is to use relabel config. For example, to filter pods we could generate:
+```
+relabel_configs:
+ - source_labels: [__meta_kubernetes_pod_node_name]
+   separator: ;
+   regex: $NODE_NAME
+   replacement: $1
+   action: keep
+```
+This could be considered as “last-mile filtering”. In other words, we watch for all the changes, and then filter the changes that match.
+
+The second option is to use field selector. For example, to select pods on a specific node, we can do:
+```
+kubernetes_sd_configs:
+   - role: pod
+ 	selectors:
+ 	- role: pod
+   	  field: spec.nodeName=$NODE_NAME
+```
+This option performs better, because we filter targets right at discovery time, and also because Kubernetes API server watch cache indexes pods by node name (as we can see in [Kubernetes codebase](https://github.com/kubernetes/kubernetes/blob/v1.30.0-rc.0/pkg/registry/core/pod/storage/storage.go#L91)). So we’ll go with this option.
+
+
+## 7. Action Plan
+
+For the implementation, we’ll do what we detailed in the How section. The common logics between StatefulSet and DaemonSet modes will be extracted into one place. We will have a separate `daemonset.go` for the separate logic of the DaemonSet mode.
+
+For the test, we will have unit tests covering new logic, and integration tests covering the basic user cases, which are:
```

Comment:
Right now the e2e tests are based on the [default kind config](https://github.com/prometheus-operator/prometheus-operator/blob/main/test/e2e/kind-conf.yaml) which deploys only one node. It isn't quite enough to test the daemonset scenarios.
I suggest to modify the configuration and deploy a 2 nodes cluster instead (provided that it doesn't break the existing e2e tests).

https://kind.sigs.k8s.io/docs/user/configuration/#nodes

Reply:
☝️ 

Let's mention that in the proposal :)